rand_distr/src/dirichlet.rs

```diff
@@ -9,11 +9,160 @@
 
 //! The dirichlet distribution.
 #![cfg(feature = "alloc")]
-use num_traits::Float;
-use crate::{Distribution, Exp1, Gamma, Open01, StandardNormal};
-use rand::Rng;
-use core::fmt;
+use crate::{Beta, Distribution, Exp1, Gamma, Open01, StandardNormal};
 use alloc::{boxed::Box, vec, vec::Vec};
+use core::fmt;
+use num_traits::{Float, NumCast};
+use rand::Rng;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+struct DirichletFromGamma<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    samplers: Box<[Gamma<F>]>,
+}
+
+impl<F> DirichletFromGamma<F>
+where
+    F: Float,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    // Construct a new `Dirichlet` with the given alpha parameter `alpha`.
+    //
+    // This function is part of a private implementation detail.
+    // It assumes that the input is correct, so no validation is done.
+    #[inline]
+    fn new(alpha: &[F]) -> DirichletFromGamma<F> {
+        let gamma_dists = alpha
+            .iter()
+            .map(|a| Gamma::new(*a, F::one()).unwrap())
```

Comment:
The `unwrap` is a bit problematic here, we probably want to propagate the error as a `Result` instead.

Reply:
Done.